Component Diagram changes are animated

sashahonchar opened the pull request:
https://user-images.githubusercontent.com/17301016/103247895-2d6d2280-49b4-11eb-9cd0-db701af7be0a.mp4

sashahonchar committed:
WIP

sashahonchar committed:
WIP2

sashahonchar committed:
Component Diagram changes are animated

dustinbyrne commented on src/js/components/componentDiagram/graph/index.js:
```diff
@@ -0,0 +1,308 @@
+        edge.group.move(edge.points);
+      } else {
+        const edgeGroup = new EdgeGroup(edge.points, this.options.animation);
+        edgeGroup.element.classList.add(`nodes---${v}---${w}`);
```
This won't work for things like HTTP routes which have a space in the name.
![image](https://user-images.githubusercontent.com/8737782/103304623-1c81d880-49d7-11eb-8c21-e59b6a0b988a.png)

dustinbyrne replied:
Also, what is the value of putting the node identifiers in the class list? How are they being used?

sashahonchar replied:
Those node names are used to emit `edge` event on Component Diagram.

dustinbyrne commented on src/js/components/componentDiagram/graph/util.js:
```diff
@@ -0,0 +1,97 @@
+  node.label = node.label || node.id;
+  node.class = node.class || node.type;
+
+  if (node.id === 'HTTP') {
+    node.label = 'HTTP server requests';
+    node.shape = 'http';
+    node.class = 'http';
```
Missing the case for children of `HTTP`, e.g. `GET /users`.

sashahonchar replied:
In this function I don't know is node a child of some special node (HTTP, SQL or package) or not.
To display HTTP children nodes with the same styles as parent 'HTTP server requests' node I'm checking the parent node ID here: 
```
https://github.com/applandinc/diagrams-js/pull/57/files#diff-a12813e59cc6d9ff69ab2da8f58567785e286a886ddd4bfd68d942e216bcda71R57-R60
```

dustinbyrne replied:
It looks like they're styled properly now

sashahonchar replied:
Seems another changes resolved this issue.

dustinbyrne commented on src/js/components/componentDiagram/graph/groups/clusterGroup.js:
```diff
@@ -0,0 +1,51 @@
+
+    setElementPosition(this, node.x, node.y);
+
+    this.element.appendChild(Rect(node.width, node.height));
```
I'm not seeing any toggle to make these cluster rects visible. Maybe I'm missing something?

sashahonchar replied:
Seems I broke it at some point. Fixed in [`93d0e55`](https://github.com/applandinc/diagrams-js/pull/57/commits/93d0e55636fdbedf9b9237a529611097869a5b86)

dustinbyrne commented on src/js/components/componentDiagram/graph/groups/nodeGroup.js:
```diff
@@ -0,0 +1,74 @@
+
+    setElementPosition(this, node.x, node.y);
+
+    this.animationOptions = animationOptions;
```
This constructor is being called with `animationOptions` as the second parameter. See:
https://github.com/applandinc/diagrams-js/pull/57/files#diff-a12813e59cc6d9ff69ab2da8f58567785e286a886ddd4bfd68d942e216bcda71R112

dustinbyrne replied:
In fact, `parent` is unused.

sashahonchar replied:
Fixed

dustinbyrne commented on src/js/components/componentDiagram/graph/index.js:
```diff
@@ -0,0 +1,308 @@
+};
+
+export default class Graph {
+  constructor(element, options = {}) {
```
The `ComponentDiagram` doesn't pass any `options`, so it seems like there's no way of changing the animation settings.

sashahonchar replied:
Pass an options and change animation settings in [`096b5a9`](https://github.com/applandinc/diagrams-js/pull/57/commits/096b5a9882ca7271b069e3f0e802bee595f83245)

dustinbyrne commented on src/js/components/componentDiagram/graph/index.js:
```diff
@@ -0,0 +1,308 @@
+        node.group = clusterGroup;
+        this.clustersGroup.appendChild(clusterGroup.element);
+      } else {
+        const nodeGroup = new NodeGroup(node, this.options.animation);
+
+        node.group = nodeGroup;
+        node.element = nodeGroup.element;
+
+        this.nodesGroup.appendChild(nodeGroup.element);
```
New nodes and edges don't get a chance to animate. I think it'd be most intuitive if the nodes spawned stacked on their parent (or somewhere amidst their children) and tweened to their final position.

Ideally, nothing ever instantly snaps into position or changes visibility.

dustinbyrne commented on src/js/components/componentDiagram/graph/util.js:
```diff
@@ -0,0 +1,97 @@
+  return result;
+}
+
+export function normalizePoints(points1, points2) {
```
Can you add a comment to explain what this is? To me, the signature reads like some sort of vector normalization.

sashahonchar replied:
Added the comment: 
`when we need to transition between 2 edges with different points length - fill smaller array with copies of it's first point`

dustinbyrne commented on src/js/components/componentDiagram/graph/util.js:
```diff
@@ -0,0 +1,97 @@
+  return el;
+}
+
+export function transformPoints(points) {
```
Can you add a comment to explain what this is?

sashahonchar replied:
Added comment: `transform 'dagre' points into simple array which will be passed to 'd3-line' helper`

dustinbyrne commented on src/js/components/componentDiagram/graph/index.js:
```diff
@@ -0,0 +1,308 @@
+    this.render();
+  }
+
+  scrollToNodes(container, nodes) {
```
UX note:

It was unclear to me why my viewport would move and zoom occasionally when expanding packages. With the movement of the viewport along with the movement of the nodes it's easy to lose track of what's going on. What if this method only translates the view and we drop the scaling?

sashahonchar replied:
In some cases when user expands the package which contains many classes he might not understand what's going on. I've recorded 2 videos:

- First - with current behaviour
- Second - without zooming out when package expands

If you want not to zoom out then user expands the package - I'll disable zooming.

https://user-images.githubusercontent.com/17301016/103816546-e0263a80-50b0-11eb-871b-88239c48bee5.mov


https://user-images.githubusercontent.com/17301016/103816560-e4525800-50b0-11eb-9947-c0a9d838ba42.mov

dustinbyrne replied:
The first video you linked does look better. Maybe what I'm seeing is a bug. If you expand a package while zoomed out it sometimes zooms in strangely.
![Jan-07-2021 12-24-15](https://user-images.githubusercontent.com/8737782/103923659-5ba0df80-50e3-11eb-99bc-28d21199cb28.gif)

dustinbyrne commented on src/js/components/componentDiagram/graph/index.js:
```diff
@@ -0,0 +1,308 @@
+const DEFAULT_OPTIONS = {
+  animation: {
+    enable: true,
+    duration: 300,
```
Somewhere between 500-1000ms feels better to me, but this may better be resolved if we were able to override this.

sashahonchar replied:
Indeed, I made 300ms as default animation speed. But you can set your own duration value like in [`096b5a9`](https://github.com/applandinc/diagrams-js/pull/57/commits/096b5a9882ca7271b069e3f0e802bee595f83245)

dustinbyrne reviewed (changes requested):
Hi @sashahonchar, this is looking great. I've added a couple notes. In addition, it'd be wonderful to test some of this behavior, at least testing for the presence of the elements we expect (for example, missing HTTP child edges).

I noticed you didn't use d3. Out of curiosity, why not? Anyway, I think the code looks fine without.

Also, could you please update the README where appropriate?

sashahonchar committed:
Remove unused NodeGroup parameter, store connecting node names in edge dataset

sashahonchar committed:
Add comments for some helpers from Graph's util.js

sashahonchar committed:
Show border of expanded package

sashahonchar committed:
Remove unused functions

sashahonchar committed:
Edit script on example page

sashahonchar committed:
Show expanded packages with fade in effect

dustinbyrne reviewed (changes requested):
The package boundaries look to be always visible. I'm OK with this, but they don't reposition correctly as other packages are expanded.

See the top package in this screenshot:
![image](https://user-images.githubusercontent.com/8737782/103924602-44aebd00-50e4-11eb-9f61-f27a51b61d60.png)

Floating package:
![image](https://user-images.githubusercontent.com/8737782/103924741-79bb0f80-50e4-11eb-84a7-733582ac3567.png)

sashahonchar committed:
Resize cluster after every graph re-render

sashahonchar commented:
Fixed an issue with cluster floating border after expanding/collapsing many packages @dustinbyrne

dustinbyrne committed:
Migrate tests to Cypress/Jest

Cypress feels more appropriate given the amount of DOM manipulation
we're doing. JSDOM will still be in use for unit testing under Jest.

dustinbyrne committed:
Fix d3 import

dustinbyrne committed:
Update dist artifacts

dustinbyrne committed:
Update CI to use Cypress

sashahonchar committed:
Update package version and changelog
